pkg/bus/bus.go

```diff
@@ -31,10 +32,13 @@ func (eb *EventBus) Publish(e data.Event) {
 	eb.rw.RLock()
 
 	for _, rf := range eb.subscribers {
-		go func(rf EventReceiveFunc) {
+		eb.wg.Add(1)
+		go func(rf EventReceiveFunc, wg *sync.WaitGroup) {
+			defer wg.Done()
```

Comment:
Doesn't this essentially eliminate any efficiency gained by running the EventReceiveFunc() as a goroutine?

Reply:
Yes and this is the reason I'm concerned about performance.

**Other ways to solve this would be:** 
Count the logs in the log handler plugin (assign an increasing number to each log) and then sort the logs in loki application plugin, which looks to me like a really bad solution.

Just create a third bus for logs / events, that need this kind of synchronization

Make this code conditional and add a configuration option for application plugins (or somewhere else) and if that option is specified for any of the plugins subscribed to the bus, then this code is used, without it we don't wait for anything.


Now that I think about it, I probably like the last option the most (maybe even more than this)

Reply:
@pleimer @paramite what do you think?

Reply:
That last option might be a good way to at least unblock logging with Loki for now, and so this patch doesn't interfere with https://github.com/infrawatch/sg-core/pull/67.

In the long run I think we will need to revisit this bus architecture and optimize/enforce order a little better. That might take some refactoring of the core.

Reply:
It should only block functions that are writing to the same subscriber, however. Even though we don't run multiple application plugins in practice, if we did, it's important that writes to one not be blocked by writes to another.

Reply:
I agree with making this conditional upon some flag that we would only supply in the sg pods that send to loki. I don't think enforcing order is important for most systems, and in fact it would be nice to find out if it's actually possible to relax this in loki (I have some guesses as to perhaps why not, but it's worth a look, I guess)

Reply:
> That last option might be a good way to at least unblock logging with Loki for now, and so this patch doesn't interfere with #67.

Can we just merge #67 ASAP or is it waiting on anything else? (302 Redirect responses to that PR please)

Reply:
As far as I know, this and #67 are totally independend. So you don't need to take this PR into account when merging #67. Even if there is some conflict, which I don't see, I plan to make this code conditional as we agreed, so I'll solve any issue when working on that.

@csibbitt There currently isn't any way to relax this in Loki. The requirement is there because of the way Loki stores data (I think). This is currently probably the most often discussed issue in the Loki community and there are some plans to remove this requirement for a little decrease in performance, but I don't know when this feature gets implemented (it could easily be a year).

Reply:
I kinda expected to hear that. Thanks for looking into it!

Reply:
The optimization done in #67 depended on posts to ES happening (basically) asynchronously. Blocking goroutines would reduce the effect of that optimization, that's what I mean by interfering with that PR.

@csibbitt it's all been approved, I think Martin was doing more testing on that, not sure what the results are

Reply:
Another option here is to re-timestamp logs according to arrival time to loki app. Question is if the differences between rsyslog timestamp and sg-core timestamp would be a problem?